### PR TITLE
Use dependencies through Gradle

### PR DESCRIPTION
**Pull request by askogvold**
Due to problems with JOGL, I downgraded it to 2.1.4

Also added:
Build complete distribution zip (including run scripts), now made the default task.
Upgraded to Gradle 2.8

I have not tested Jython yet, and I have not upgraded the run scripts.
Also, when the dependencies are gone from the repo, I suppose the Ant build would need to go as well.

Any pointers to more that I should test before this may be considered complete?

### CODE REVIEW

**Comment by dhale** [comment]
I like this change. Although the dependency list is relatively small, the Mines JTK should manage dependencies in a more standard way than me occasionally updating jars in our repo. When I have done that in the past, I was careful to run all of the tests, especially those in lapack, ogl, and sgl. The lapack unit tests are automated, but I had to manually observe and manipulate the graphics displayed by the ogl and sgl tests. To test the jython jar I would just run some of the demos written in Jython. All of this for Mac OS X, Windows, and Ubuntu Linux.

Can you briefly summarize what in the Mines JTK is broken when using JOGL 2.1.5?
(This is the version I tested over a year ago. Did I miss something?)

Has this been fixed in a later version? 
(The current version of JOGL is 2.3.2.)

On what platforms have you tested your changes?

How does the new default task affect the default build time?

To others: anyone still using Ant to build the Mines JTK?

**Comment by askogvold** [comment]
Regarding JOGL versions, I had a problem compiling with Gradle, as it
reports a Debug.java in JOGL to be duplicate. That did not seem to be a
problem in newer versions, but they had then removed a bunch of deprecated
methods, and moved from the javax namespace to jogamp.

There are necessary changes if we are to upgrade JOGL, and the Gl class
wrapping GL2 needs some attention. If you want to have a look, you could
just change the version for gluegen and jogl in build.gradle to the latest.

The dist task is performed after the jar task, so it just collects and zips
the built jar and dependencies.

I have not run all the tests yet, but will have another go at it over the
weekend. I made this PR a bit prematurely in order to get some early
feedback.

tor. 22. okt. 2015, 17.56 skrev Dave Hale <notifications@github.com>:

**Comment by askogvold** [comment]
Running gradlew test:
edu.mines.jtk.util.UnitsTest > testDefine FAILED
    junit.framework.AssertionFailedError at UnitsTest.java:35
280 tests completed, 1 failed

When I run only that single test, it does not fail. This seems to be caused by the Units singleton already being defined, so that adding Fahrenheit returns false because it's already defined.

For the 3D tests, I'm not sure what to look for, but I now see that I get an exception because gluegen-rt.so is not in the correct path. Will look into this.

**Comment by chrisengelsma** [comment]
@askogvold I am encountering the same issue re: UnitsTest, been unable to reproduce it predictably.

**Comment by chrisengelsma** [comment]
Also, regarding the duplicates issue, I start running into that issue, but throwing this into the build.gradle file:

```
jar {
...
  duplicatesStrategy = 'exclude'
...
}
```

fixes it for me. Not sure if this a good permanent solution or not

**Comment by askogvold** [comment]
Updated to JOGL 2.3.2. I found the script for generating the Gl class, and ran it on the newest documentation. A few constants reported ambiguity between GL2ES1 and GL2ES2. They had the same value, so I just referenced the GL2ES1 version instead of GL2 where there were ambiguities.

All tests run green now, the application compiles nicely, and the native dependencies are resolved correctly in runtime. ( ./gradlew test )

Ran all the main methods in sgl and ogl, and they seem to make sense.

Tested jython ( ./jythonrun <full path to python script> ), and it seems to work.

Tested with IntelliJ IDEA on Linux x64 (RHEL 6.7).

Have not updated Windows launch scripts.

**Comment by askogvold** [comment]
@chrisengelsma : The duplication issue disappeared with the newer JOGL 2.3.2.

**Comment by dhale** [comment]
Anyone want to test on Mac OS and Windows?

Especially the 3D graphics tests and demos in ogl and sgl, because of the different native code libraries required.

**Comment by dhale** [comment]
By the way, Arve's changes are an outstanding upgrade. Well worth testing on other platforms.

**Comment by askogvold** [comment]
Updated windows bat files, and verified that they are working as well.

The usage instructions may also be simplified with the changes from this branch. Now the user can either extract the zip file in 'build/distributions', or it could be provided through 'gradlew installDist', which will put the extracted application in build/install/

I rebased my branch in order to get your latest changes. Hope that did not mess things up too much. You might need to delete your local version of this branch, and then fetch it again.

**Comment by dhale** [comment]
Both javarun.sh and jythonrun.sh fail on Mac OS, which does not support the -f option to readlink. 

More generally, it is unclear how the new scripts and their Windows equivalents should be used. Are you assuming that users will add .../jtk/bin/ to their PATH? (Not recommended.) The current scripts have hardwired pathnames, but others can easily modify them (and other parameters) for their own systems. 
As discussed in the README, these scripts should be examples, and I do not know how to use them.

**Comment by askogvold** [comment]
Ok, sorry about that - I don't have a Mac and haven't tested it there.

SCRIPTPATH is evaluated to the path where the run script is located, and then the CLASSPATH is set relative from there to the jar files in the same folder and in the lib folder.

readlink is used to make this work also for symlinks to the script, and I think the -f is to allow recursion through several symlinks in order to get the actual path. The -f option could be removed and it would work in most (if not all) use cases.

Usage:
1. gradlew distZip (or just gradlew, it's the default task) - another option is distTar
2. extract the archive where you want it
3. call the executable script you want (in the folder where the application was unpacked) -- you do not need to stand in the path of the script.

When you're using an IDE, you just run the classes you wish within the project, and things should work out of the box.

**Comment by dhale** [comment]
I have updated our file layout and build to be consistent with modern conventions, dependency management, and so on.

Note especially the change from edu_mines_jtk.jar to edu-mines-jtk-1.0.0.jar.

Arve, would you please check to see if these changes enable us to close this issue? Your pull request inspired much of this work --- you will recognize your Gl.java --- and our discussion here about scripts for running Mines JTK programs led me to now favor the use of Gradle for that purpose.

Just this morning I published edu-mines-jtk-1.0.0.jar (with javadoc and sources) to Maven Central. We should soon be able to search Maven Central and find this:
https://repo1.maven.org/maven2/edu/mines/jtk/edu-mines-jtk/1.0.0/
In other words, anyone can now use the Mines Java Toolkit without first building it themselves.

**Comment by askogvold** [comment]
Hi

I have not run or tested your newest version, but this matches the goal with my change very well -- jtk available through one specified dependency in maven central. Also good to see the new run scripts through Gradle, which makes the usage very straight-forward. 

Very good!

My next step after merge would be to split the project in logical parts -- so that users do not need to depend on the 3D graphics package if they're not using it. I suppose you have the best overview of the interdependencies, and which modules the project would consist of, but I can have a look at this as well if you want.

Feel free to close this PR, I think you did all of this and more :rocket:

**Comment by dhale** [comment]
I am inclined to wait for Java 9's modules before splitting further (beyond core and demo). Meanwhile, I will close this PR. 

A search for "mines" in Maven Central now yields our latest version:
http://search.maven.org/#artifactdetails%7Cedu.mines.jtk%7Cedu-mines-jtk%7C1.0.0%7Cjar
